### PR TITLE
feat: introduce CommunitiesKeyDistributor

### PR DESCRIPTION
**Pull request by osmaczko**
Introduced `CommunitiesKeyDistributor`

This component decouples key distribution from the Messenger, enhancing code maintainability, extensibility and testability. It also alleviates the need to impact all methods potentially affecting encryption keys.
Moreover, it allows key distribution inspection for integration tests.

part of: status-im/status-desktop#10998

### CODE REVIEW

File: protocol/messenger.go [hunk 1]
```diff
@@ -460,19 +461,23 @@ func NewMessenger(
...
+		transport:              transp,
+		encryptor:              encryptionProtocol,
+		sender:                 sender,
+		anonMetricsClient:      anonMetricsClient,
+		anonMetricsServer:      anonMetricsServer,
+		telemetryClient:        telemetryClient,
+		pushNotificationClient: pushNotificationClient,
+		pushNotificationServer: pushNotificationServer,
+		communitiesManager:     communitiesManager,
+		communitiesKeyDistributor: &CommunitiesKeyDistributorImpl{
```
Comment (0x-r4bbit) [inline protocol/messenger.go:488]: Just cosmetics, but we could have a `NewKeyDistributor()` factory fn for this

File: protocol/messenger_communities.go [hunk 2]
```diff
@@ -211,11 +211,45 @@ func (m *Messenger) handleCommunitiesHistoryArchivesSubscription(c chan *communi
...
+			m.logger.Warn("failed to retrieve orgs", zap.Error(err))
+			return result
+		}
+
+		for _, org := range ownedOrgs {
+			result[org.IDString()] = org
+		}
+
+		return result
+	}()
```
Comment (0x-r4bbit) [inline protocol/messenger_communities.go:231]: Any reason this is wrapped in an IIFE? Nothing that needs change, just curious

**Review by 0x-r4bbit — APPROVED** [review]

File: protocol/communities/community.go [hunk 3]
```diff
@@ -1363,6 +1363,7 @@ func (o *Community) AddTokenPermission(permission *protobuf.CommunityTokenPermis
 	}
 
 	o.config.CommunityDescription.TokenPermissions[permission.Id] = permission
+	o.updateEncrypted()
```
Comment (mprakhov) [inline protocol/communities/community.go]: If I understood correctly, if you'll do updateEcrypted, `UpdateCommunityEncryption` will always return nil due to
```
if community.Encrypted() == isEncrypted {
		return nil
	}
```

Comment (mprakhov) [reply protocol/communities/community.go]: ok, I see, it was removed

Comment (mprakhov) [reply protocol/communities/community.go:1366]: What about admin events? What if the admin will add member permission? Add/Delete/Update token CommunityEvent must include encryption information in this case

File: protocol/messenger_communities.go (hunk 2 above)
Comment (osmaczko) [reply protocol/messenger_communities.go:231]: It represents the intention a bit better, you can read straight away "this piece of code is responsible for obtaining recentlyPublishedOrgs". Also, this way I don't pollute the scope with unneeded `ownedOrgs` variable.

File: protocol/messenger.go (hunk 1 above)
Comment (osmaczko) [reply protocol/messenger.go:488]: Was thinking about factory as well, but since this struct doesn't need any special initialization, I left it as is.

**Review by jrainville — APPROVED** [review]

**Comment by osmaczko** [comment]
Waits for QA: https://github.com/status-im/status-desktop/pull/11657

File: protocol/common/raw_message.go [hunk 4]
```diff
@@ -35,4 +35,5 @@ type RawMessage struct {
 	CommunityKeyExMsgType CommKeyExMsgType
 	Ephemeral             bool
 	BeforeDispatch        func(*RawMessage) error
+	HashRatchetGroupID    []byte
```
Comment (Samyoul) [inline protocol/common/raw_message.go:38]: This is a nice change.

File: protocol/communities_messenger_test.go [hunk 5]
```diff
@@ -3447,47 +3447,49 @@ func (s *MessengerCommunitiesSuite) TestStartCommunityRekeyLoop() {
 	s.Require().False(c.Encrypted())
 	// TODO some check that there are no keys for the community. Alt for s.Require().Zero(c.RekeyedAt().Unix())
 
-	// Update the community to use encryption and check the values
```
Comment (Samyoul) [inline protocol/communities_messenger_test.go:3453]: were your new changes causing this test to fail?

**Review by Samyoul — APPROVED** [review]
Thank you for your patience and your work here with this PR. 🙏

File: protocol/communities_messenger_test.go (hunk 5 above)
Comment (osmaczko) [reply protocol/communities_messenger_test.go:3453]: It didn't compile because I removed `UpdateCommunityEncryption`. I will address that in follow-up.

**Pull request merged**